src/core/profiling/endoscope_backend.c

```diff
@@ -0,0 +1,577 @@
+  /* thread safe applied */
+  GRPC_ENDO_INDEX atom_index;
+  gpr_mu_lock(&(base->mutex));
+  if (base->atom_stack == GRPC_ENDO_EMPTY) {
+    grpc_endo_recycle_atoms(base);
+    if (base->atom_stack == GRPC_ENDO_EMPTY) {  /* still empty, no atom gained by recycling */
+      gpr_mu_unlock(&(base->mutex));
+      grpc_endo_warning_print("grpc_endo_create_atom: no atom item available or recyclable to allocate");
+      grpc_endo_warning_span(base, GRPC_ENDO_WARNING_ATOM);
+      return GRPC_ENDO_EMPTY;
+    }
+  }
+  atom_index = base->atom_stack;
+  base->atom_stack = base->atom_pool[atom_index].next_atom;
+  gpr_mu_unlock(&(base->mutex));
+  base->atom_pool[atom_index].next_atom = GRPC_ENDO_EMPTY;
+  return atom_index;
+}
+
+static GRPC_ENDO_INDEX grpc_endo_create_task(grpc_endo_base *base) {
+  /* thread safe applied */
+  GRPC_ENDO_INDEX task_index = GRPC_ENDO_EMPTY;
+  gpr_mu_lock(&(base->mutex));
+  if (base->task_stack != GRPC_ENDO_EMPTY) {  /* task item available */
+    task_index = base->task_stack;
+    base->task_stack = base->task_pool[task_index].next_task;
+    base->task_pool[task_index].task_id = base->task_count++;
+    gpr_mu_unlock(&(base->mutex));
+  } else if (base->task_history_head != GRPC_ENDO_EMPTY) {  /* recycle oldest history task */
+    task_index = base->task_history_head;
+    if (base->task_withatom_head == task_index) {  /* if oldest history task has atoms then recycle */
+      grpc_endo_recycle_atoms(base);
+    }
+    base->task_history_head = base->task_pool[task_index].next_task;
+    base->task_pool[task_index].task_id = base->task_count++;
+    gpr_mu_unlock(&(base->mutex));
+  } else {  /* no available and cannot recycle */
+    gpr_mu_unlock(&(base->mutex));
+    grpc_endo_warning_print("grpc_endo_create_task: no task item available or recyclable to allocate");
+    grpc_endo_warning_span(base, GRPC_ENDO_WARNING_TASK);
+    return GRPC_ENDO_EMPTY;
+  }
+  base->task_pool[task_index].next_task = GRPC_ENDO_EMPTY;
+  /* ASSERT base->task_pool[task_index].next_taskwithatom != EMPTY */
+  return task_index;
+}
+
+/* Marker Implementation */
+
+gpr_int64* grpc_endo_begin(grpc_endo_base *base, const char *name, const char *file, const gpr_int32 line,
+    const char* function_name) {  /* return the position to set for cycle_now() */
+  /* no thread safety needed assuming no conflict on the same thread */
+  GRPC_ENDO_INDEX marker_id;
+  gpr_uint64 thread_id;
+  GRPC_ENDO_INDEX thread_index;
+  grpc_endo_thread *mythread;
+  gpr_int64 cycle_created = grpc_endo_cyclenow();
+
+  marker_id = grpc_endo_get_or_create_marker(base, line, name);
+  if (marker_id == GRPC_ENDO_EMPTY) {
+    grpc_endo_warning_print("grpc_endo_begin: cannot create marker item");
+    return &grpc_endo_invalid64;  /* only when out of markers */
+  } else {
+    grpc_endo_marker *mymarker = &(base->marker_pool[marker_id]);
+    if (mymarker->name == NULL) {  /* created new marker if not found */
+#ifdef GRPC_ENDO_COPY_MARKER_NAME
```

Comment:
Let's remove `GRPC_ENDO_COPY_MARKER_NAME` and set the policy that these are never copied